### PR TITLE
t2 root

### PR DESCRIPTION
**Pull request by Student007**
@johnnyman727 please take a look at both commit comments and additional requests on slack :smile:

### CODE REVIEW

**Comment by wprater** [comment]
thanks for your submission.  I have not followed all of this, but could you add a brief description in the PR so anyone can see quickly grok the context of this PR (help for historical reasons as well)?  might I also suggest that you remove all code that's [been](https://github.com/tessel/t2-cli/pull/307/files#diff-f68567477d803b49930337bf7fe1556bR707)  [commented](https://github.com/tessel/t2-cli/pull/307/files#diff-f68567477d803b49930337bf7fe1556bR672 ) [out](https://github.com/tessel/t2-cli/pull/307/files#diff-f68567477d803b49930337bf7fe1556bR669)?  it's confusing for a code reviewer and should not be in the deployed codebase.

File: bin/tessel-2.js [hunk 1]
```diff
@@ -7,6 +7,7 @@ var key = require('../lib/key');
 var init = require('../lib/init');
 var logs = require('../lib/logs');
 
+
```
Comment (johnnyman727) [inline bin/tessel-2.js]: This is super nit-picky but before requesting reviews on PRs, can you remove whitespace changes?

File: lib/controller.js [hunk 2]
```diff
@@ -92,7 +95,7 @@ Tessel.list = function(opts) {
 
 Tessel.get = function(opts) {
   return new Promise(function(resolve, reject) {
-    logs.info('Looking for your Tessel...');
+    logs.info('Searching nearby LAN (netmask) and USB for Tessels...');
```
Comment (johnnyman727) [inline lib/controller.js]: This message is confusing - can you change it back to the original?

File: lib/controller.js [hunk 3]
```diff
@@ -7,7 +7,10 @@ var sprintf = require('sprintf-js').sprintf;
 var cp = require('child_process');
 var async = require('async');
 var updates = require('./update-fetch');
+var debug = require('debug')('controller');
+var Menu = require('terminal-menu');
 var controller = {};
+controller.TESSEL_AUTH_KEY = Tessel.TESSEL_AUTH_PATH + '/id_rsa';
```
Comment (johnnyman727) [inline lib/controller.js]: I commented on the other PR, but I think this variable should actually be defined and exported in `provision.js` since that file is responsible for writing the keys. What do you think?

File: lib/controller.js [hunk 4]
```diff
@@ -641,6 +663,213 @@ controller.tesselFirmwareVerion = function(opts) {
...
+    if (tessel.lanConnection) {
+      if (!tessel.lanConnection.authorized) {
+        rtm.add(index + ') ' + tessel.name + ': ' + tessel.lanConnection.ip + ' (not authorized) \n');
+      } else {
+        rtm.add(index + ') ' + tessel.name + ': ' + tessel.lanConnection.ip + ' (authorized) \n');
+      }
+
+    } else if (tessel.usbConnection) {
+      if (!tessel.usbConnection.authorized) {
+        rtm.add(index + ') ' + tessel.name + ': [USB] (not authorized) \n');
```
Comment (johnnyman727) [inline lib/controller.js]: USB Connections are always authorized

File: lib/controller.js [hunk 5]
```diff
@@ -641,6 +663,213 @@ controller.tesselFirmwareVerion = function(opts) {
...
+  a ncurses like menu you can select the Tessel you like to gain root access...
+
+  The structure of code and maybe unusual parts are paid due being testable.
+  The lightweight terminal-menu isn't written testable what needs a little bit hacking.
+
+  Finally some parts are causing from my personnal learning curve about writing unit tests using sinon!
+*/
+controller.root = function(opts) {
+  // ~ conversion to home because spawn isn't able to handle this right
+  if (opts.path && opts.path.substring(0, 1) === '~') {
```
Comment (johnnyman727) [inline lib/controller.js]: Could you just use the [`osenv`](https://github.com/tessel/t2-cli/blob/master/lib/tessel/provision.js#L12) module here?

File: lib/controller.js [hunk 6]
```diff
@@ -641,6 +663,213 @@ controller.tesselFirmwareVerion = function(opts) {
...
+    var home = process.env[(process.platform === 'win32') ? 'USERPROFILE' : 'HOME'];
+    if (/^~/.test(opts.path)) {
+      opts.path = opts.path.replace(/~/, home);
+    }
+  }
+  opts.root = true;
+  return new Promise(function(resolve, reject) {
+    controller.menu.seek(opts)
```
Comment (johnnyman727) [inline lib/controller.js]: This should be something like:
```.js
// Grab the preferred Tessel
  return Tessel.get(opts)
    .then(function(tessel) {
      // Open up an SSH session
      return tessel.beginSSHSession(opts)
        .then(function() {
          return controller.closeTesselConnections(tessel);
        });
    });
```

As far as I can tell, `controller.menu.seek` is the same as `Tessel.get` so we might as well make it consistent with all the other commands.

File: lib/controller.js [hunk 7]
```diff
@@ -641,6 +663,213 @@ controller.tesselFirmwareVerion = function(opts) {
         });
     });
 };
+controller.menu = {
```
Comment (johnnyman727) [inline lib/controller.js]: Let's remove all `menu`-related code and put it in it's own PR. It's not specifically related to SSH sessions.

File: lib/controller.js [hunk 8]
```diff
@@ -181,8 +204,8 @@ controller.closeTesselConnections = function(tessels) {
   return new Promise(function(resolve, reject) {
     async.each(tessels, function closeThem(tessel, done) {
         // If not an unauthorized LAN Tessel, it's connected
-        if (!(tessel.connection.connectionType === 'LAN' &&
-            !tessel.connection.authorized)) {
+        if (!(tessel.lanConnection &&
```
Comment (johnnyman727) [inline lib/controller.js]: Good catch!

File: lib/controller.js [hunk 9]
```diff
@@ -150,15 +166,22 @@ Tessel.get = function(opts) {
...
-        connectionsToClose.splice(tessels.indexOf(tessel), 1);
+        if (opts.root) {
 
-        controller.closeTesselConnections(connectionsToClose)
-          .then(function() {
-            return resolve(tessel);
-          });
+          controller.closeTesselConnections(connectionsToClose)
+            .then(function() {
+              controller.ssh.runSSH(0, opts, tessels, resolve, reject);
```
Comment (johnnyman727) [inline lib/controller.js]: @Student007, you shouldn't need to modify the `Tessel.get` function. It should be a black box that returns Tessels to functions that ask for it and shouldn't have any references to any other specific Tessel commands. You should call the `ssh` function from within `controller.root` after requesting the Tessel from `Tessel.get`.

File: lib/controller.js (hunk 6 above)
Comment (johnnyman727) [inline lib/controller.js]: Additionally, if you need to make sure that you get a `lan` Tessel, you can set the `--lan` flag to true.

File: lib/controller.js [hunk 10]
```diff
@@ -641,6 +663,213 @@ controller.tesselFirmwareVerion = function(opts) {
...
+  },
+  notAuthorized: function() {
+    logs.warn('Sorry, you are not authorized!');
+    logs.info('Maybe this Tessel didn\'t got "t2 provision" until now?');
+  }
+};
+
+/*
+  The T2 root command is used to login into the firmware and gaining superuser access.
+  The security is provided by RSA - your Tessel get the keys while provisioning via USB.
```
Comment (johnnyman727) [inline lib/controller.js]: typo - 'your Tessel get the keys'. I suggest you make it 'your Tessel receives the keys over USB when provisioned'.

File: lib/controller.js [hunk 11]
```diff
@@ -641,6 +663,213 @@ controller.tesselFirmwareVerion = function(opts) {
...
+  
+  If you have only one Tessel, the T2 root command will login directly else there is
```
Comment (johnnyman727) [inline lib/controller.js]: I suggest you remove this. Being able to use the menu should be something you can do for any Tessel command.

File: lib/controller.js [hunk 12]
```diff
@@ -641,6 +663,213 @@ controller.tesselFirmwareVerion = function(opts) {
...
+};
+
+/*
+  The T2 root command is used to login into the firmware and gaining superuser access.
+  The security is provided by RSA - your Tessel get the keys while provisioning via USB.
+  
+  If you have only one Tessel, the T2 root command will login directly else there is 
+  a ncurses like menu you can select the Tessel you like to gain root access...
+
+  The structure of code and maybe unusual parts are paid due being testable.
```
Comment (johnnyman727) [inline lib/controller.js]: Remove this - we'll make it work but it shouldn't be on the code in production.

File: lib/controller.js [hunk 13]
```diff
@@ -641,6 +663,213 @@ controller.tesselFirmwareVerion = function(opts) {
...
+  The lightweight terminal-menu isn't written testable what needs a little bit hacking.
+
+  Finally some parts are causing from my personnal learning curve about writing unit tests using sinon!
```
Comment (johnnyman727) [inline lib/controller.js]: Remove this as well.

File: lib/controller.js [hunk 14]
```diff
@@ -641,6 +663,213 @@ controller.tesselFirmwareVerion = function(opts) {
...
+  },
+  exit: function(resolve) {
+    resolve();
+    //process.exit();
+  }
+};
+controller.ssh = {
+  runSSH: function(id, opts, tessels, resolve, reject) {
+    // clear console
+    debug('runSSH...\n root@' + tessels[id].lanConnection.host);
```
Comment (johnnyman727) [inline lib/controller.js]: This throws an error if a Tessel with only a USB connection is provided. You need to check for that and reject with an error if it's not possible.

File: lib/controller.js [hunk 15]
```diff
@@ -641,6 +663,213 @@ controller.tesselFirmwareVerion = function(opts) {
...
+      var ch = require('child_process')
+        .spawn('ssh', ['-i',
+          opts.path,
+          'root@' + tessels[id].lanConnection.ip
+        ], {
+          stdio: 'inherit'
+        });
+      logs.info('Connect to ' + tessels[id].lanConnection.host + '...');
+
+      // FIXME: There is no handler for poweroff Tessel ! Needs to be fixed within the firmware...
```
Comment (johnnyman727) [inline lib/controller.js]: What exactly are you looking for here? A notification that Tessel is turning off?

File: lib/controller.js [hunk 16]
```diff
@@ -641,6 +663,213 @@ controller.tesselFirmwareVerion = function(opts) {
...
+  exit: function(resolve) {
+    resolve();
+    //process.exit();
+  }
+};
+controller.ssh = {
+  runSSH: function(id, opts, tessels, resolve, reject) {
+    // clear console
+    debug('runSSH...\n root@' + tessels[id].lanConnection.host);
+    if (opts.menu) {
```
Comment (johnnyman727) [inline lib/controller.js]: What is the menu option for?

File: lib/controller.js [hunk 17]
```diff
@@ -641,6 +663,213 @@ controller.tesselFirmwareVerion = function(opts) {
...
+        });
+      logs.info('Connect to ' + tessels[id].lanConnection.host + '...');
+
+      // FIXME: There is no handler for poweroff Tessel ! Needs to be fixed within the firmware...
+      // (Terminal freezes imediatelly after poweroff the Tessel)
+
+      ch.once('error', function(e) {
+        if (e === 255) {
+          controller.menu.clear();
+          controller.ssh.notAuthorized();
```
Comment (johnnyman727) [inline lib/controller.js]: How would it get to this point if it's not authorized?

**Comment by johnnyman727** [comment]
In addition to @wprater's valid comments, I've added some suggestions to the code as well. Thanks for contributing @Student007!

File: bin/tessel-2.js (hunk 1 above)
Comment (wprater) [inline bin/tessel-2.js]: I think you can write a rule for this in jscs.  so it will fail travis builds.. or are you only using jshint

**Comment by Student007** [comment]
@wprater sorry - I was in a harry and fretful of someone will change things while I am preparing my pull request. As you can see ... there are conflicts wasn't there yesterday.

**Comment by wprater** [comment]
you can rebase from master and fix them.  they are probably from the
changes I made today.  should not be so tough to fix?

File: lib/controller.js (hunk 2 above)
Comment (Student007) [inline lib/controller.js]: @johnnyman727 with the change of possible multiple `Looking for your Tessels ...` but being earnest: I think it is more useful to get an info about where and what is searched... 
The old version suggests the CLI is created for a single Tessel - and looking to the code I feel like this had bin the initial purpose ?!

File: lib/controller.js (hunk 3 above)
Comment (Student007) [inline lib/controller.js]: - [Tessel.TESSEL_AUTH_PATH + '/id_rsa'](https://github.com/Student007/t2-cli/blob/master/lib/lan_connection.js#L32) -> used in lan_connection.js
- [path.join(os.tmpDir(),'id_rsa')](https://github.com/ericvicenti/ssh-keygen/blob/master/src/ssh-keygen.js#L98) -> extern ssh-keygen module
- [filepath](https://github.com/Student007/t2-cli/blob/master/lib/tessel/provision.js#L3-L13) -> provision.js

`Tessel` is included in `provision.js` and also `lan_connection.js` but useless in `bin/tessel-2.js`!
The only solution would be to use `controller.js` to provide that *const* because `controller.js` is included within `bin/tessel-2.js`. A possible solution could be:
```
controller.TESSEL_AUTH_KEY = Tessel.TESSEL_AUTH_PATH+'/id_rsa';
```
within [controller.js](https://github.com/Student007/t2-cli/blob/master/lib/controller.js#L10)

For this reason I decided to create `controller.TESSEL_AUTH_KEY`

Comment (Student007) [inline lib/controller.js]: it is used in `bin/tessel-2.js` to replace `default: default_id_rsa` due to your request for reusable central.

File: lib/controller.js (hunk 4 above)
Comment (Student007) [inline lib/controller.js]: @johnnyman727 - no. If I use a Tessel is not provisioned at this computer, I may not login. If this isn't the case, it is a big security problem and requires to open a major issue bound to milestone

File: lib/controller.js (hunk 5 above)
Comment (Student007) [inline lib/controller.js]: @johnnyman727 you could add this comment also [in answer to @rwaldron wishes](https://github.com/tessel/t2-cli/pull/290/files#r39067831).
How do you think the replacement should look like ?

File: lib/controller.js (hunk 6 above)
Comment (Student007) [inline lib/controller.js]: @johnnyman727 as `controller.menu.seek` calls `Tessel.get` - the reason is `unit testing ...` 
(but of cause I can kick the tests and do understandable code - it causes from the internals of `terminal-menu` code desing)

About the `--lan` flag ? What do you mean ? The `t2 root` command doesn't know that.

File: lib/controller.js (hunk 4 above)
Comment (Student007) [inline lib/controller.js]: I do not understand the security concept ? Could you explain please ?

File: lib/controller.js (hunk 7 above)
Comment (Student007) [inline lib/controller.js]: https://github.com/tessel/t2-cli/pull/309 ok

File: lib/controller.js (hunk 10 above)
Comment (Student007) [inline lib/controller.js]: ok - but why not `Tessels receive their keys over USB when provisioned` - I am in conflict with thinking about only one Tessel.

File: lib/controller.js (hunk 11 above)
Comment (Student007) [inline lib/controller.js]: you mean `If you have only one Tessel, the T2 root command will login directly ` ?

File: lib/controller.js (hunk 12 above)
Comment (Student007) [inline lib/controller.js]: see

File: lib/controller.js (hunk 13 above)
Comment (Student007) [inline lib/controller.js]: yes .. nobody should know or remember it was hard to dive into Tessel development ;)

File: lib/controller.js (hunk 14 above)
Comment (Student007) [inline lib/controller.js]: yes, yes ... good to know people own a Tessel 2 .... ... ...

File: lib/controller.js (hunk 15 above)
Comment (Student007) [inline lib/controller.js]: I think it is not possible to fix this problem. The terminal freeze if you power down the Tessel 2 virtual machine. If you do `poweroff` logged in to a Tessel 2 by `t2 root` (ssh) this behavior is bad ... but there is no easy solution. I would expect it is fixing poweroff command by looking for open terminals and exit them with error code before power down. The comment is due to opening an issue or make a central note about knowledge about this behavior.

File: lib/controller.js (hunk 16 above)
Comment (Student007) [inline lib/controller.js]: it is a testing switch

File: lib/controller.js (hunk 17 above)
Comment (Student007) [inline lib/controller.js]: the `t2 root` command isn't interested in whether a Tessel 2 is authorized. It would be able to use a different `id_rsa` by `t2 root -i http://local/key_from_cloud` (e.g.) and login to a Tessel. In the case the key is wrong ... this error code is responded. 

At this point: Once more ... the solution with a flag "authorized" isn't helpful for creating a menu or do other things like replacing id_rsa ...

I think at the moment I need to create
```
controller.ssh = {
// ....
  checkAuthorized = function(ip,id_rsa){
   var ch = require('child_process')
        .spawn('ssh', ['-i',
         id_rsa,
          'root@' + ip
        ], {
          stdio: 'inherit'
        });
    ch.once('error', function(e) {
        if (e === 255) {
          return false;
       }
      });
    ch.once('exit', function(e) {
     if (e === 255) {
      return false;   
      } else if(e === 0){
      return true;
    }
 });
ch.kill('SIGHUP');
  }
}

```
And call it within 
```
controller.menu = {
// ...
 makeMenu: function(tessel, index, rtm) {
    if (tessel.lanConnection) {
var auth = controller.ssh.checkAuthorized(tessel.lanConnection.ip,opts.path)
// if (!tessel.lanConnection.authorized) {
if (!auth) {
rtm.add(index + ') ' + tessel.name + ': ' + tessel.lanConnection.ip + ' (not authorized) \n');
      } else {
        rtm.add(index + ') ' + tessel.name + ': ' + tessel.lanConnection.ip + ' (authorized) \n');
      }
// ...
```

Comment (johnnyman727) [inline lib/controller.js]: Oh I understand - so this is if a user provides a key file but once the SSH process runs, T2 decides the keys aren't legit?

Comment (Student007) [inline lib/controller.js]: If I use the `t2-cli`, I will do this multiple times - each time for a single job. And if I (would... ...) own multiple Tessels, it is possible I like to have multiple keys (one for each Tessel). So if I run a command like `t2 root` or `t2 list`, I will use `-i key_of_tesselXYZ`or `-i key_of_Customer123s_Tessels` ...(e.g.). Based on this, the design of being authorized should be reviewed. 

I think I didn't got it all together in my mind (especial about USB is never `not authorized`) - but I feel something isn't perfect at this point.

File: bin/tessel-2.js (hunk 1 above)
Comment (rwaldron) [inline bin/tessel-2.js]: https://github.com/tessel/t2-cli/blob/master/.jscsrc

But I don't think there is a rule that says "no new empty lines"

File: lib/controller.js (hunk 2 above)
Comment (rwaldron) [inline lib/controller.js]: > I think it is more useful to get an info about where and what is searched... 

Any tessel-centric command allows for `--usb` or `--lan` flags, so this message won't make sense if one of those is used.

File: bin/tessel-2.js (hunk 1 above)
Comment (Student007) [inline bin/tessel-2.js]: I will do :+1:

File: lib/controller.js (hunk 2 above)
Comment (Student007) [inline lib/controller.js]: @rwaldron didn't notice this until I added the current experimental `t2 select` command. But in this case this could be a switch around `opts.lan`:
```
if(opts.lan) {
  logs.info('Searching nearby LAN (netmask) for Tessels...');
} else if(opts.usb) {
  logs.info('Searching USB for your Tessel...');
} else {
  logs.info('Searching nearby LAN (netmask) and USB for Tessels...');
} 
```
**Question**: Because I do not own a single Tessel 2 .. ... . I don't know whether `var seeker = new Seeker ..` also discovers USB Tessels (I assume) ... but does it differ between `--lan`and `--usb` or is this done by filtering later ?

File: lib/controller.js (hunk 16 above)
Comment (Student007) [inline lib/controller.js]: @johnnyman727 do you agree to remove testing flags and only write tests for things what are not blowing up code by lines like this ? Also the stupid testing related organizing code parts like `controller.ssh.seek` are only necessary for testing and looks bad for someone don't write tests... In fact this feels stupid and terminal-menu integration isn't really testable but I think the normal case would be to write unit tests for terminal-menu itself.
What do you prefer ?
@rwaldron maybe your meaning is interesting here (did you wrote the most of the tests - or who has this done (Jon?) ?)

File: lib/controller.js (hunk 6 above)
Comment (Student007) [inline lib/controller.js]: @johnnyman727 got it while writing `t2 select` ;)

**Comment by johnnyman727** [comment]
@Student007 I was away for the weekend but will try to provide more feedback tomorrow. Still digging through emails.

File: lib/controller.js (hunk 2 above)
Comment (johnnyman727) [inline lib/controller.js]: @Student007 I get where you are coming from but 90% of users won't care what types of connections the CLI is looking for. They will just want to deploy their code and it should work regardless of the connection type. Let's remove it from the PR because it's not related to `t2 root` and if you feel strongly about it, let's open an issue so more people can discuss. :)

File: lib/controller.js (hunk 3 above)
Comment (johnnyman727) [inline lib/controller.js]: I think TESSEL_AUTH_KEY and TESSEL_AUTH_PATH should be set in the same place. For someone looking to maintain this code in the future, it would be confusing that one is set in `controller.js` and the other is set in `provision.js`. 

@Student007 I think the better option would be to not set the default in `tessel-2.js` and just substitute the standard Tessel key path in `controller.js`. 

I'm always partially thinking about how we can make this CLI eventually usable for other types of devices and to do that we would need to separate the device key location from the command parsing.

File: lib/controller.js (hunk 4 above)
Comment (johnnyman727) [inline lib/controller.js]: @Student007 with Tessel 2, we presume that if it's connected over USB, then that user has full authorization to use it. All they have to do is run `t2 provision` to now have control over it even if it's not plugged in via USB.

File: lib/controller.js (hunk 10 above)
Comment (johnnyman727) [inline lib/controller.js]: @Student007 that works too! I just want to make sure the grammar is correct.

File: lib/controller.js (hunk 11 above)
Comment (johnnyman727) [inline lib/controller.js]: Yep! Just like if you only have one Tessel and you do `t2 run`, it will automatically deploy to that Tessel. It provides for the best user experience because most people will only have one Tessel.

File: lib/controller.js (hunk 15 above)
Comment (johnnyman727) [inline lib/controller.js]: Ah well, we'll need to address this before this PR can be merged. I'm surprised the child process doesn't try to quit after the SSH session is merged...

File: lib/controller.js (hunk 16 above)
Comment (johnnyman727) [inline lib/controller.js]: @Student007 I would prefer to impact the code as little as possible with unit test modifications. You should be able to create stubs for any of these methods within your unit tests to change the behavior however you need. [Here](https://github.com/tessel/t2-cli/blob/master/test/unit/deploy.js#L238-L242) is an example for how to do that.

File: lib/controller.js (hunk 15 above)
Comment (IgitDanny) [inline lib/controller.js]: @johnnyman727 it would be fixed by using `t2 reboot` what is an [other pull request](https://github.com/tessel/t2-cli/pull/328)

File: lib/controller.js (hunk 17 above)
Comment (johnnyman727) [inline lib/controller.js]: @Student007 our model for T2 only involves one SSH key pair for all of your Tessels. I'm not sure what the point of having multiple keys would be.

I do understand that if someone passes in an alternate key path to this particular command then it's possible that it won't be correct and the command will fail.

You'll want to set the Tessel.TESSEL_AUTH_PATH with the provided key before calling `Tessel.get` so that we attempt to use that key to open an SSH connection with any discovered Tessels. That way you will know by this point whether the keys work or not.

**Comment by johnnyman727** [comment]
@Student007 I went through and added more feedback! Let me know if you need help rebasing or making suggested changes.

File: lib/controller.js (hunk 2 above)
Comment (Student007) [inline lib/controller.js]: @johnnyman727 as mentioned I opened an issue https://github.com/tessel/t2-cli/issues/340 for further discussions.

File: lib/controller.js (hunk 3 above)
Comment (Student007) [inline lib/controller.js]: > I'm always partially thinking about how we can make this CLI eventually usable for other types of devices and to do that we would need to separate the device key location from the command parsing.

@johnnyman727  That is a good idea. In that case we should have a second and third VM to be discovered. That way we can hold this in mind.

By writing this, I got awash by TODOs like
- replace all "Tessel" by DEVICETYPENAME
- TESSEL_AUTH_PATH should become '.'+DEVICETYPENAME+AUTH_PATH
- ...++

What type of devices ? Tessel based or also Ra. PI, Intel an others ?
Or do you mean to be able to fork and easy re-engineering this CLI for e.g. Ra. PI for running OpenWRT on it ?

File: lib/controller.js (hunk 4 above)
Comment (Student007) [inline lib/controller.js]: @johnnyman727 OK, thanks, now I understand the security idea. A single Tessel is bond to a single computer after provisioning. I hear ringing the :bell: and see a high frequented FAQ entry:
# Tessel 2 FAQ
> ### My Tessel doesn't work on my second Computer
> For reasons of security your Tessel is only able to `run` and `push` your scripts from a single computer - that one you use `t2 provision`. No other computers are able to connect to your Tessel. But you are able to connect also by LAN from that single computer so you are able to `push` scripts through the air :ghost: :v: 
*-You are visitor 1999*

:see_no_evil: :hear_no_evil: :speak_no_evil: :wink: only my two cents ...

But this is not the worst idea :+1: 

Maybe it would be an idea to provide a solution about using a key-server what accepts sURL based authorization for putting generated keys and use them remotely.

---
Btw: I didn't got this until now, because .... I ... didn't ... own ... a single Tessel 2 ...

File: lib/controller.js (hunk 15 above)
Comment (Student007) [inline lib/controller.js]: @johnnyman727 I will do a new PR but will use this one for discussion. I am waiting to prepare a new PR until the PR for `controller.menu`, `TESSEL_AUTH_KEY-option` and the local branch about `t2 select` is ready and able to merge and are merged - due to the parts I like to use from that PRs.

File: lib/controller.js (hunk 3 above)
Comment (johnnyman727) [inline lib/controller.js]: Yes, devices like RasPi and Edison which also run some version of Linux. It will definitely take some refactoring but it would be an interesting goal.

File: lib/controller.js (hunk 4 above)
Comment (johnnyman727) [inline lib/controller.js]: Yes, it's possible some folks will be confused about that. We were initially planning on making `t2 provision` run automatically but ultimately decided against it so that users don't get confused about trying to deploy to LAN devices that they've never connected over USB before.

File: lib/controller.js (hunk 5 above)
Comment (johnnyman727) [inline lib/controller.js]: I suppose we should break out that logic into it's own functionality that can be called both in `deploy.js` and here.

File: lib/controller.js (hunk 3 above)
Comment (Student007) [inline lib/controller.js]: :+1: I will have this in mind - but I will ignore my current wishes to look over everything at this point - due to you would wait much longer to merge a useful result causing from discussion and feedback :smile:

File: lib/controller.js (hunk 4 above)
Comment (Student007) [inline lib/controller.js]: ok - thanks for talking about backgrounds and internal history.

File: lib/controller.js (hunk 17 above)
Comment (Student007) [inline lib/controller.js]: > @Student007 our model for T2 only involves one SSH key pair for all of your Tessels. I'm not sure what the point of having multiple keys would be.

@johnnyman727 I don't like this because it would be the same as not being able to change PASSWORDs... 
I vote for having `--key` options and review the code for necessary changes as I have done as first step in https://github.com/tessel/t2-cli/pull/330 
Also if you gave the key to someone or got hacked whatever ... **changing and having different keys for different Tessels** should be possible.
My two cents ...
Opened an issue for further discussion https://github.com/tessel/t2-cli/issues/347

Comment (johnnyman727) [inline lib/controller.js]: Thanks for making a separate issue. That will allow us to discuss in more detail with more people and merge PRs in the meantime.

**Comment by johnnyman727** [comment]
@Student007 what else do we need to get this PR merged?

**Comment by johnnyman727** [comment]
Closing in favor of #453.
